src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs

```diff
@@ -5,130 +5,188 @@
 using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
 using NuGet.Common;
 using NuGet.Configuration;
+using NuGet.Protocol.Core.Types;
 using NuGet.Protocol.Utility;
 
 namespace NuGet.VisualStudio.Telemetry
 {
     public sealed class PackageSourceTelemetry : IDisposable
     {
         private readonly ConcurrentDictionary<string, Data> _data;
-        private readonly IDictionary<string, PackageSource> _sources;
+        private readonly IDictionary<string, SourceRepository> _sources;
         private readonly Guid _parentId;
+        private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, string>> _resourceStringTable;
+        private readonly string _actionName;
+        private readonly HashSet<string> _knownSources;
```

Comment:
Do you still need _knownSources? Looks like it's covered by _sources.

Reply:
`_knownSources` was intended for performance, the `ProtocolDiagnosticEvent` events will be called a large number of times, so the difference between a linear lookup for a `List` could be much slower than a quick `HashSet<>.Contains`.  However, from another comment realized that if I pre-populated the `_data`, then it can double as a fast known-source whitelist, as well as a dictionary of the data type.